tsdb/engine/tsm1/cache.go

```diff
@@ -138,6 +139,14 @@ func (c *Cache) Snapshot() *Cache {
 	snapshot := NewCache(c.maxSize)
 	snapshot.store = c.store
 	snapshot.size = c.size
+	snapshot.dirty = make(map[string]*entry)
```

Comment:
@jwilder I don't think I fully understand your suggestion. Are you suggesting we don't need snapshot.dirty?  If not, could you elaborate your suggestion a with an example?

Assuming you are asking about the need  for a separate `snapshot.dirty` map.

We need to keep the old `c.store` because we need to keep what was in the cache live to readers. We use a new map `snapshot.dirty` because we need a map that is visible only to the snapshotter which it is doing its deduplication work (which means we don't need to lock it).
 
`snapshot.dirty` keeps the snapshotter and all other readers of the cache out of each others way for the duration of the deduplication.